lib/block_view_association.go

```diff
@@ -753,83 +797,132 @@ func (bav *UtxoView) GetUserAssociationsByAttributes(associationQuery *UserAssoc
+		break
+	}
+	if associationQuery.SortDescending {
+		// Sort descending, so add UTXO view association entries first.
+		for _, associationEntry := range utxoViewAssociationEntries {
+			if associationQuery.Limit > 0 && len(associationEntries) >= associationQuery.Limit {
+				break
+			}
+			associationEntries = append(associationEntries, associationEntry)
+		}
+		// Then add database association entries.
+		for _, associationEntry := range dbAssociationEntries {
+			if associationQuery.Limit > 0 && len(associationEntries) >= associationQuery.Limit {
+				break
+			}
+			if _, exists := utxoViewAssociationEntryMap[associationEntry.AssociationID]; !exists {
+				associationEntries = append(associationEntries, associationEntry)
+			}
+		}
+	} else {
+		// Track which association entries we've already added from
+		// the UTXO view so as not to duplicate returned results.
+		addedAssociationEntryIDs := NewSet[*BlockHash]([]*BlockHash{})
+
+		// Sort ascending, so add database association entries first.
+		for _, associationEntry := range dbAssociationEntries {
+			if associationQuery.Limit > 0 && len(associationEntries) >= associationQuery.Limit {
+				break
+			}
+			if utxoViewAssociationEntry, exists := utxoViewAssociationEntryMap[associationEntry.AssociationID]; exists {
+				addedAssociationEntryIDs.Add(utxoViewAssociationEntry.AssociationID)
+				associationEntries = append(associationEntries, utxoViewAssociationEntry)
+			} else {
+				addedAssociationEntryIDs.Add(associationEntry.AssociationID)
+				associationEntries = append(associationEntries, associationEntry)
+			}
+		}
+		// Then add UTXO view association entries.
+		for _, associationEntry := range utxoViewAssociationEntries {
+			if associationQuery.Limit > 0 && len(associationEntries) >= associationQuery.Limit {
+				break
+			}
+			if !addedAssociationEntryIDs.Includes(associationEntry.AssociationID) {
+				associationEntries = append(associationEntries, associationEntry)
+			}
+		}
 	}
 	return associationEntries, nil
 }
 
 func (bav *UtxoView) CountUserAssociationsByAttributes(associationQuery *UserAssociationQuery) (uint64, error) {
 	// Validate query params.
-	err := _isValidUserAssociationQuery(associationQuery)
+	err := _isValidCountUserAssociationQuery(associationQuery)
 	if err != nil {
 		return 0, errors.Wrap(err, "CountUserAssociationsByAttributes: ")
 	}
+	// First pull matching association entries from the UTXO view so that
+	// we can track deleted association entries and can properly limit
+	// the number of entries retrieved from the database.
+	newAssociationEntries, deletedAssociationIdMap := bav._getUtxoViewUserAssociationEntriesByAttributes(associationQuery)
 	// Pull matching association IDs from the db.
-	associationIDs, err := bav.GetDbAdapter().GetUserAssociationIdsByAttributes(associationQuery)
+	associationIDs, _, err := bav.GetDbAdapter().GetUserAssociationIdsByAttributes(associationQuery, deletedAssociationIdMap)
```

Comment:
okay so kinda crazyyyyy idea. but really what the second param to `GetUserAssociationIdsByAttributes` is really just a set of `AssociationIds` to skip. What if we just add all the `associationIds` from `newAssociationEntries` to the `deletedAssociationIdMap` before we pass it in? then on the other side, we can really just return `len(newAssociationEntries) + len(associationIds)`. Thoughts? and we change the lingo from `deleted` to `excludeKeys` all the way down the stack to `EnumerateKeysforPrefixWithLimitOffsetOrder`.

Reply:
this applies to `GetUserAssociationsByAttributes` too btw